components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,409 @@
+    const auto* spender_address = ca_dict->FindString(kSpenderAddress);
+    const auto* amount = ca_dict->FindString(kAmount);
+
+    if (!approver_address || !contract_address || !spender_address || !amount ||
+        base::ToUpperASCII(*approver_address) !=
+            base::ToUpperASCII(hex_account_address)) {
+      continue;
+    }
+
+    allowance_map.insert_or_assign(
+        GetAllowanceMapKey(*contract_address, *approver_address,
+                           *spender_address),
+        mojom::AllowanceInfo::New(chain_id, *contract_address,
+                                  *approver_address, *spender_address,
+                                  *amount));
+  }
+}
+
+// Processes each data portion retrieved from logs
+void EthAllowanceManager::OnGetAllowances(
+    const int& task_id,
+    const std::string& chain_id,
+    const std::string& hex_account_address,
+    const std::vector<Log>& logs,
+    base::Value rawlogs,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::ProviderError::kSuccess) {
+    if (allowance_discovery_tasks_.find(task_id) !=
+        allowance_discovery_tasks_.end()) {
+      allowance_discovery_tasks_[task_id]->MarkComplete();
+      MergeAllResultsAnCallBack();
+    }
+    return;
+  }
+
+  std::vector<Log> logs_tmp{logs};
+  std::sort(logs_tmp.begin(), logs_tmp.end(),
+            [](const Log& first, const Log& second) -> bool {
+              return std::tie(first.block_number, first.log_index) <
+                     std::tie(second.block_number, second.log_index);
+            });
+
+  // Collection of the latest allowances per contract & spender & approver.
+  std::map<std::string, mojom::AllowanceInfoPtr> allowance_map;
+  // Put cached data into the map first if available.
+  uint256_t latest_block_number{0};
+  LoadCachedAllowances(chain_id, hex_account_address, latest_block_number,
+                       allowance_map);
+  for (const auto& log_item : logs_tmp) {
+    // Skip pending logs.
+    if (log_item.block_number == 0) {
+      continue;
+    }
+
+    if (log_item.topics.size() < kTopicElementsCountToCheck) {
+      continue;
+    }
+    const auto current_map_key = GetAllowanceMapKey(
+        log_item.address, log_item.topics[1], log_item.topics[2]);
+
+    uint256_t parsed_amount{0};
+    if (!HexValueToUint256(log_item.data, &parsed_amount)) {
+      continue;
+    }
+
+    if (parsed_amount > 0) {
+      // Replace if same key exists by the fresh allowance data.
+      allowance_map.insert_or_assign(
+          current_map_key, mojom::AllowanceInfo::New(
+                               chain_id, log_item.address, log_item.topics[1],
+                               log_item.topics[2], log_item.data));
+    } else if (allowance_map.find(current_map_key) != allowance_map.end()) {
+      allowance_map.erase(current_map_key);
+    }
+    latest_block_number = log_item.block_number;
```

Comment:
latest_block_number should become the second one of 
```
    filter_options.Set(kFromBlock, get_blocknumber_filter_from_cache(
                                       chain_id, account_address_hex));
    filter_options.Set(kToBlock, latest_block_num_hex);
```
So when next run of discover happens this cached value+1 is used as a value for kFromBlock.

We don't need to look at `log_item.block_number` at all.

Reply:
fixed